polaris-core/src/main/java/org/apache/polaris/core/persistence/BasePersistence.java

```diff
@@ -0,0 +1,341 @@
+      @Nonnull PolarisEntityType entityType,
+      int limit,
+      @Nonnull Predicate<PolarisBaseEntity> entityFilter,
+      @Nonnull Function<PolarisBaseEntity, T> transformer);
+
+  /**
+   * Lookup the current entityGrantRecordsVersion for the specified entity. That version is changed
+   * everytime a grant record is added or removed on a base securable or added to a grantee.
+   *
+   * @param callCtx call context
+   * @param catalogId catalog id or NULL_ID
+   * @param entityId unique entity id
+   * @return current grant records version for that entity.
+   */
+  int lookupEntityGrantRecordsVersion(
+      @Nonnull PolarisCallContext callCtx, long catalogId, long entityId);
+
+  /**
+   * Lookup the specified grant record from the grant_records table. Return NULL if not found
+   *
+   * @param callCtx call context
+   * @param securableCatalogId catalog id of the securable entity, NULL_ID if the entity is
+   *     top-level
+   * @param securableId id of the securable entity
+   * @param granteeCatalogId catalog id of the grantee entity, NULL_ID if the entity is top-level
+   * @param granteeId id of the grantee entity
+   * @param privilegeCode code for the privilege we are looking up
+   * @return the grant record if found, NULL if not found
+   */
+  @Nullable
+  PolarisGrantRecord lookupGrantRecord(
+      @Nonnull PolarisCallContext callCtx,
+      long securableCatalogId,
+      long securableId,
+      long granteeCatalogId,
+      long granteeId,
+      int privilegeCode);
+
+  /**
+   * Get all grant records on the specified securable entity.
+   *
+   * @param callCtx call context
+   * @param securableCatalogId catalog id of the securable entity, NULL_ID if the entity is
+   *     top-level
+   * @param securableId id of the securable entity
+   * @return the list of grant records for the specified securable
+   */
+  @Nonnull
+  List<PolarisGrantRecord> loadAllGrantRecordsOnSecurable(
+      @Nonnull PolarisCallContext callCtx, long securableCatalogId, long securableId);
+
+  /**
+   * Get all grant records granted to the specified grantee entity.
+   *
+   * @param callCtx call context
+   * @param granteeCatalogId catalog id of the grantee entity, NULL_ID if the entity is top-level
+   * @param granteeId id of the grantee entity
+   * @return the list of grant records for the specified grantee
+   */
+  @Nonnull
+  List<PolarisGrantRecord> loadAllGrantRecordsOnGrantee(
+      @Nonnull PolarisCallContext callCtx, long granteeCatalogId, long granteeId);
+
+  /**
+   * Check if the specified parent entity has children.
+   *
+   * @param callContext the polaris call context
+   * @param optionalEntityType if not null, only check for the specified type, else check for all
+   *     types of children entities
+   * @param catalogId id of the catalog
+   * @param parentId id of the parent, either a namespace or a catalog
+   * @return true if the parent entity has children
+   */
+  boolean hasChildren(
```

Comment:
Why is it critical to have this as a dedicated API method?

Reply:
Good question. Not sure, for now just preserving behavior. Looks like there's a mix of using `hasChildren` and `listEntities(limit...)`

Reply:
If you do not mind, let's remove it to reduce the API breadth. We can always add something later if this use case needs to be optimized... I kind of doubt that it will even come to that because deleting namespace is less frequent than getting the children of a namespace.